### ISSUE TITLE
question: pinger example

### ISSUE THREAD

**Issue by unixunion** [knalli/angular-vertxbus#60]
Hi,

I am having trouble understanding why the controller doesn't connect on page load, since my vertx connection is in a unknown state. can you please provide more information on how the connection is initialized? 

a hello world example using the vert.x ping-address would be great!

/Kegan

**Comment by knalli** [comment]
There is a demo included, please have a look at the readme.

**Comment by unixunion** [comment]
I have the following in my controller:

```
angular.module('myApp.eventbus', ['knalli.angular-vertxbus'])

	.config(function(vertxEventBusProvider) {
    vertxEventBusProvider
      .enable()
      .useReconnect()
      .useUrlServer('http://localhost:8080');
  })

  .controller('EventbusCtrl', function(vertxEventBus, vertxEventBusService) {
  	console.log("eventbus controller");
    // using the EventBus directly
    vertxEventBus.send('ping-address', {data: 123});

    // using the service
    vertxEventBusService.send('ping-address', {data: 123})

  });
```

and the following in my HTML

```
  <div ng-controller="EventbusCtrl as eventbusCtrl"></div>
```

and I get the following error:

```
Error: INVALID_STATE_ERR
    at checkOpen (index.js:199)
    at sendOrPub (index.js:184)
    at vertx.EventBus.that.send (index.js:67)
    at EventbusWrapper.send (angular-vertxbus.js:1434)
    at new <anonymous> (eventbus.js:13)
    at Object.invoke (angular.js:4443)
    at $get.extend.instance (angular.js:8991)
    at nodeLinkFn (angular.js:8101)
    at compositeLinkFn (angular.js:7534)
    at compositeLinkFn (angular.js:7538)
```

no connection attempts to my vert.x instance. So I thing I am missing some call to actually start the controller?

**Comment by knalli** [comment]
When your controller is being inited, the vertxbus is still not connected yet. You should wait for the `'vertx-eventbus.system.connected'` event.

**Comment by knalli** [comment]
Alternatively, use the service.

**Comment by unixunion** [comment]
Sorry, do you have an example of using `service`? or listening for the 'vertx-eventbus.system.connected' event?  I'm not sure where / when `service` is available or how to instantiate the controller like that?

**Comment by unixunion** [comment]
I have tried this:

```
angular.module('myApp.eventbus', ['knalli.angular-vertxbus'])

	.config(function(vertxEventBusProvider) {
    vertxEventBusProvider
      .enable()
      .useReconnect()
      .useUrlServer('http://localhost:8080');
  })

  .controller('EventbusCtrl', function($scope) {
  	console.log("eventbus controller");

  	$scope.$on('vertx-eventbus.system.connected', function (event, data) {
  		console.log("connected event"); // 'Data to send'
  		console.log(event);
  		console.log(data);
		});

  });

```

and HTML

```
  <div ng-controller="EventbusCtrl as eventbusCtrl" class="ng-scope"></div>
```

but my on-connect event never seems to fire. Sorry, I know I'm doing something stupid.

**Comment by unixunion** [comment]
I am expecting vertx-eventbus to connect on page load.

**Comment by knalli** [comment]
You should use `$rootScope.$on(event, handler)`. See https://knalli.github.io/angular-vertxbus.docs/docs/#/api/knalli.angular-vertxbus.vertxEventBusService#events

**Comment by unixunion** [comment]
I still don't seem to get the event. 

```
angular.module('myApp.eventbus', ['knalli.angular-vertxbus'])
	
	.config(function(vertxEventBusProvider) {
    vertxEventBusProvider
      .enable()
      .useReconnect()
      .useUrlServer('http://localhost:8080');
  })

  .controller('EventbusCtrl', function($scope, $rootScope) {
  	console.log("eventbus controller init");

  	$rootScope.$on('vertx-eventbus.system.connected', function (event, data) {
  		console.log("eventbus connected"); // 'Data to send'
  		console.log(event);
  		console.log(data);
		});

		$rootScope.$on('vertx-eventbus.system.disconnected', function (event, data) {
			console.log("eventbus disconnected");
  		console.log(event);
  		console.log(data);
		});

  })

```

does the connection initialise automatically if the controller is included in the project? I get the log entry
`eventbus controller init` so I know its calling init, but it doesn't seem to attempt any connections. Do I need to implement a .factory?

**Comment by knalli** [comment]
No, that's fine. I would suggest looking into the WebSocket connection. It looks like it is not working (in Chrome you can see all frames in live).

Again: Did you have taken a look into the demo? This one should work, so you can compare your local setup..

**Comment by unixunion** [comment]
I have based my code off your example, and I get connected now finally! but the registerHandler's do NOT seem to fire on-connect, they are triggered from click events in your Demo app. 

How do I get the registerHandler's to connect on start and reRegister on reconnect?

**Comment by unixunion** [comment]
I figured it out. I was trying to create a pointer in $scope but this worked.

```

   $rootScope.$on('vertx-eventbus.system.connected', function (event, data) {
      console.log('Vert.X connected (status)', data);
      $rootScope.sessionIsValid = (data && data.status === 'ok');
      vertxEventBusService.on('broadcast', function (data) {
        console.log("received broadcast: " + data);
        $rootScope.$broadcast("broadcast", { msg: data });
      });
    });

```

**Comment by unixunion** [comment]
I am receiving duplicate messages from the vertx instance. Is there currently an issue with Vert.x 3? 

I am definitely only running one instance of the subscriber, and it is receiving the message twice from the vert.x 3 server component.

**Comment by unixunion** [comment]
server:

```
    vertx.setPeriodic(1000, tping -> {
      eb.publish("ping", new JsonObject().put("body", "ping"));
    });
```

client:

```
'use strict';

  angular.module('myApp.eventbus', ['ng', 'knalli.angular-vertxbus'])
  .config(function(vertxEventBusProvider) {
    vertxEventBusProvider
    .useDebug(true)
    .useUrlServer('http://localhost:8080');
  })
  .run(function ($rootScope, vertxEventBus, vertxEventBusService, $interval) {
    console.log("eventbus run method");
    $rootScope.sessionIsValid = false;
    $rootScope.$on('vertx-eventbus.system.login.succeeded', function (event, data) {
      console.log('Vert.X Login succeeded (status)', data);
      $rootScope.sessionIsValid = (data && data.status === 'ok');
    });
    $rootScope.$on('vertx-eventbus.system.login.failed', function (event, data) {
      console.log('Vert.X Login failed (status)', data);
      $rootScope.sessionIsValid = false;
    });
    $rootScope.$on('vertx-eventbus.system.connected', function (event, data) {
      console.log('Vert.X connected (status)', data);
      $rootScope.sessionIsValid = (data && data.status === 'ok');
      vertxEventBusService.on('ping', function (data) {
        // Broadcast.handle(data);
        console.log("eventbus received broadcast: " + data.body);
        $rootScope.$broadcast("notifications", { msg: data });
      });
    });
    $rootScope.$on('vertx-eventbus.system.disconnected', function (event, data) {
      console.log('Vert.X disconnected (status)', data);
      $rootScope.sessionIsValid = false;
    });

    $rootScope.moduleStats = {
      wrapper: {},
      service: {}
    };
    $interval(function () {
      try {
        $rootScope.moduleStats.wrapper.readyState = vertxEventBus.readyState();
        $rootScope.moduleStats.service.readyState = vertxEventBusService.readyState();
        $rootScope.moduleStats.service.getConnectionState = vertxEventBusService.getConnectionState();
        $rootScope.moduleStats.service.isEnabled = vertxEventBusService.isEnabled();
        $rootScope.moduleStats.service.isConnected = vertxEventBusService.isConnected();
      } catch (e) {}
    }, 1000);
  })
  .filter('eventBusState', function () {
    console.log(".filter method");
    var states = ['CONNECTING', 'OPEN', 'CLOSING', 'CLOSED'];
    return function (value) {
      return states[value] || value;
    };
  })
  .controller('EventbusCtrl', function($scope, vertxEventBusService) {
    console.log(".controller method");
    $scope.login = function (username, password) {
      vertxEventBusService.login(username, password)
        .then(function (reply) {
          $scope.reply = reply;
        })
        ['catch'](function (reply) {
          $scope.reply = reply;
        });
    };
    var holder = {};
  });
```


the receive listener for "ping" fires twice on every message the server emits.

```
[Vert.x EB Stub] Enabled
angular.js:12314 [Vert.x EB Stub] Enabled: connecting 'http://localhost:8080/eventbus'
eventbus.js:10 eventbus run method
app.js:14 app run
angular.js:12314 [Vert.x EB Stub] Connected
eventbus.js:21 Vert.X connected (status) undefined
eventbus.js:25 eventbus received broadcast: ping1
notifications.js:10 notification handler
eventbus.js:25 eventbus received broadcast: ping1
notifications.js:10 notification handler
eventbus.js:25 eventbus received broadcast: ping2
notifications.js:10 notification handler
eventbus.js:25 eventbus received broadcast: ping2
```

**Comment by unixunion** [comment]
It seems like the low-level vertx client is registering twice,  I added some debug statements to vertx-angular and the low level vertxbus.js modules. I get 1 register event in the `angular-vertxbus` module, but it seems to call twice in the low level driver? 

```
[Vert.x EB Stub] Enabled
angular.js:12314 [Vert.x EB Stub] Enabled: connecting 'http://localhost:8080/eventbus'
eventbus.js:10 eventbus run method
app.js:14 app run
angular.js:12314 [Vert.x EB Stub] Connected
eventbus.js:21 Vert.X connected (status)[object Object] data undefined
angular-vertxbus/dist/angular-vertxbus.js:275 vertx-angular dist register handler
vertxbus.js/index.js:75 vertx registerHandler
vertxbus.js/index.js:75 vertx registerHandler
```

**Comment by unixunion** [comment]
If I comment out line ~252 in angular-vertxbus.js the duplicate registration issue goes away.

the line:
```
              this.delegate.registerHandler(address, callback);
```

**Comment by knalli** [comment]
> I figured it out. I was trying to create a pointer in $scope but this worked.

Pointer? If you mean event listener: [Yes](https://github.com/knalli/angular-vertxbus/issues/60#issuecomment-113515198) :)

> ```
    $rootScope.$on('vertx-eventbus.system.connected', function (event, data) {
      console.log('Vert.X connected (status)', data);
      $rootScope.sessionIsValid = (data && data.status === 'ok');
      vertxEventBusService.on('broadcast', function (data) {
        console.log("received broadcast: " + data);
        $rootScope.$broadcast("broadcast", { msg: data });
      });
    });
```

Actually, you don't have to register the listener (via service) after the connected event (because it is passive).

> I am receiving duplicate messages from the vertx instance. Is there currently an issue with Vert.x 3?

Sounds strange, although, it's still not finally tested against vertx-v3.

Do you have this issue with the demo included also?

**Comment by unixunion** [comment]
can be reproduced with the demo, but ONLY if you register the handlers on `system.connected` event in `.run`, then you get duplicate messages.


add the following to the `.run` root scope and remove the registration in `.controller` to reproduce.
```

    $rootScope.$on('vertx-eventbus.system.connected', function(event, data) {
      var holder = {};
        holder.timeServiceDeconstructor = vertxEventBusService.on('what-time-is-it', function (data) {
          console.log("what time is it2 ");
        });
    });

```

**Comment by knalli** [comment]
Thanks!

Fixed in f3311ca5, use the latest 2.0.4

**Comment by knalli** [comment]
Also I've updated the demo code, there had been some issues I was not aware of.
